src/components/tl/ucp/scatterv/scatterv_linear.c

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "scatterv.h"
+#include "core/ucc_progress_queue.h"
+#include "tl_ucp_sendrecv.h"
+
+static inline ucc_rank_t get_peer(ucc_rank_t rank, ucc_rank_t size,
+                                  ucc_rank_t step)
+{
+    return (rank + step) % size;
+}
+
+void ucc_tl_ucp_scatterv_linear_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task  = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team  = TASK_TEAM(task);
+    ucc_coll_args_t   *args  = &TASK_ARGS(task);
+    void*              sbuf  = args->src.info_v.buffer;
+    ucc_memory_type_t  smem  = args->src.info_v.mem_type;
+    ucc_rank_t         grank = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         gsize = UCC_TL_TEAM_SIZE(team);
+    int                polls = 0;
+    ucc_rank_t         peer, posts, nreqs;
+    size_t             dt_size, data_size, data_displ;
+
+    if (UCC_IS_ROOT(*args, grank)) {
+        posts   = UCC_TL_UCP_TEAM_LIB(team)->cfg.scatterv_linear_num_posts;
+        nreqs   = (posts > gsize || posts == 0) ? gsize : posts;
+        dt_size = ucc_dt_size(TASK_ARGS(task).src.info_v.datatype);
+
+        while (polls++ < task->n_polls) {
+            while ((task->tagged.send_posted < gsize) &&
+                   ((task->tagged.send_posted - task->tagged.send_completed) <
+                    nreqs)) {
+                peer       = get_peer(grank, gsize, task->tagged.send_posted);
+                data_size  = ucc_coll_args_get_count(args,
+                                args->src.info_v.counts, peer) * dt_size;
+                data_displ = ucc_coll_args_get_displacement(args,
+                                args->src.info_v.displacements, peer) * dt_size;
+                UCPCHECK_GOTO(ucc_tl_ucp_send_nz(PTR_OFFSET(sbuf, data_displ),
+                                                 data_size, smem, peer, team, task),
+                               task, out);
+                polls = 0;
+            }
```

Comment:
i think you should have ucc_tl_ucp_test somewhere here, to make use of outer while n_polls loop. Otherwise, there is no polling here.

Reply:
right, I will add worker progress here similar to alltoall